In [2]:
pip install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121

Looking in indexes: https://download.pytorch.org/whl/nightly/cu121
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB ? eta -:--:--
     ---------------------------------------- 0.0/2.5 GB 1.2 MB/s eta 0:34:12
     ---------------------------------------- 0.0/2.5 GB 1.2 MB/s eta 0:34:12
     ---------------------------------------- 0.0/2.5 GB 1.2 MB/s eta 0:34:12
     ---------------------------------------- 0.0/2.5 GB 1.0 MB/s eta 0:40:14
     ---------------------------------------- 0.0/2.5 GB 1.3 MB/s eta 0:32:06
     ---------------------------------------- 0.0/2.5 GB 1.3 MB/s eta 0:32:06
     ----------------------------------

In [3]:
import torch  # For PyTorch
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0))

True
NVIDIA GeForce RTX 3070 Laptop GPU


In [4]:
!pip install transformers

   ---------------------------------------- 0.0/10.1 MB ? eta -:--:--
   --- ------------------------------------ 0.8/10.1 MB 5.6 MB/s eta 0:00:02
   ------- -------------------------------- 1.8/10.1 MB 5.9 MB/s eta 0:00:02
   ---------- ----------------------------- 2.6/10.1 MB 4.4 MB/s eta 0:00:02
   ------------- -------------------------- 3.4/10.1 MB 4.7 MB/s eta 0:00:02
   ----------------- ---------------------- 4.5/10.1 MB 4.6 MB/s eta 0:00:02
   --------------------- ------------------ 5.5/10.1 MB 4.8 MB/s eta 0:00:01
   ----------------------- ---------------- 6.0/10.1 MB 4.5 MB/s eta 0:00:01
   ------------------------ --------------- 6.3/10.1 MB 4.4 MB/s eta 0:00:01
   --------------------------- ------------ 7.1/10.1 MB 3.9 MB/s eta 0:00:01
   --------------------------- ------------ 7.1/10.1 MB 3.9 MB/s eta 0:00:01
   ------------------------------ --------- 7.6/10.1 MB 3.4 MB/s eta 0:00:01
   ------------------------------ --------- 7.6/10.1 MB 3.4 MB/s eta 0:00:01
   ---

In [5]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Sep__8_19:56:38_Pacific_Daylight_Time_2023
Cuda compilation tools, release 12.3, V12.3.52
Build cuda_12.3.r12.3/compiler.33281558_0


In [6]:
from transformers import AutoTokenizer, AutoModelForCausalLM,pipeline

C:\Users\ghass\anaconda\envs\LLM\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [7]:
pip install git+https://github.com/huggingface/transformers

  Cloning https://github.com/huggingface/transformers to c:\users\ghass\appdata\local\temp\pip-req-build-1ocnlk8h
  Resolved https://github.com/huggingface/transformers to commit add53e25ffa3d1750a944086d2fbb016aee35406
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for transformers: filename=transformers-4.48.0.dev0-py3-none-any.whl size=10214330 sha256=d307a472bdcb0d52712c019ffb198cd521c73de01ca76cd79619ff4bc61449be
  Stored in directory: C:\Users\ghass\AppData\Local\Temp\pip-ephem-wheel-cache-jzd4pkmz\wheels\04\a3\f1\b88775f8e1665827525b19ac7590250f1038d947067beba9fb
Successfully built transformers
  Attempting uninstall: transformers
    Found existing installation: transformers 4.47.0
  

  Running command git clone --filter=blob:none --quiet https://github.com/huggingface/transformers 'C:\Users\ghass\AppData\Local\Temp\pip-req-build-1ocnlk8h'


In [1]:
import torch

In [2]:
from transformers import AutoTokenizer, AutoModelForCausalLM,pipeline

C:\Users\ghass\anaconda\envs\LLM\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [6]:
!pip install accelerate>=0.26.0

In [9]:
!pip install transformers datasets huggingface-hub   

   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/25.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/25.1 MB ? eta -:--:--
   - -------------------------------------- 0.8/25.1 MB 479.7 kB/s eta 0:00:51
   -- ------------------------------------- 1.6/25.1 MB 1.0 MB/s eta 0:00:24
   --- ------------------------------------ 2.4/25.1 MB 1.4 MB/s eta 0:00:

In [1]:
from huggingface_hub import login

# Use your Hugging Face token here
login(token="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")

C:\Users\ghass\anaconda\envs\LLM\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
from transformers import AutoModelForCausalLM, AutoTokenizer


In [5]:
import torch

In [7]:
model_name = "h2oai/h2o-danube3-4b-chat"

tokenizer = AutoTokenizer.from_pretrained(
    model_name,
)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=torch.bfloat16,
    device_map="auto",
    trust_remote_code=True,
)


Trying to resume download...
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████| 2/2 [00:15<00:00,  7.89s/it]
Some parameters are on the meta device because they were offloaded to the cpu.


In [6]:
!pip install sentencepiece


   ---------------------------------------- 0.0/991.5 kB ? eta -:--:--
   ------------------------------- -------- 786.4/991.5 kB 6.6 MB/s eta 0:00:01
   ---------------------------------------- 991.5/991.5 kB 5.8 MB/s eta 0:00:00


In [8]:
!pip install --upgrade protobuf


In [8]:
import pandas as pd

In [10]:
df = pd.read_csv("C:/Users/ghass/Downloads/1000_enron_spam_data.csv")
df

Unnamed: 0  Message ID  \
0             0        6568   
1             1        9797   
2             2         655   
3             3        8206   
4             4       25787   
..          ...         ...   
995         995       29351   
996         996       32821   
997         997       14476   
998         998       21733   
999         999       21910   

                                               Subject  \
0                           kirstee ' s role in london   
1    i never sent you this - keep it hush hush ! : ...   
2                                      copanos changes   
3                                    hrgovcic , hrvoje   
4                         a chance to get new logo now   
..                                                 ...   
995                                 can you help ! ! !   
996            dont let age deter you from great sex !   
997               proposed retention bonuses for netco   
998                epson inkjet cartridges from 5 . 99   
999    we can give you disccounted errection mads ooze   

                                               Message Spam/Ham        Date  
0    vince :\nthis is precisely the concern i have ...      ham  2000-07-25  
1    online credit breakthroughrepair your credit o...     spam  2002-08-02  
2    - - - - - - - - - - - - - - - - - - - - - - fo...      ham  2000-04-12  
3    please incease the bonus for hrgovcic in vince...      ham  2001-01-13  
4    working on your company ' s image ? start with...     spam  2005-06-27  
..                                                 ...      ...         ...  
995  from : mrs . victoria ndlovu\nattn :\nyou may ...     spam  2004-08-13  
996  20 minutes is all it takes\nmore info here . ....     spam  2005-05-13  
997  louise ,\ni am sorry to bother you at such a c...      ham  2001-12-28  
998  save up to 75 % on inkjet , laser copier suppl...     spam  2005-04-18  
999  this is one of our newest sites , thus we ' d ...     spam  2005-05-10  

[1000 rows x 6 columns]

In [12]:
df = df[:120]

In [13]:
df

Unnamed: 0  Message ID  \
0             0        6568   
1             1        9797   
2             2         655   
3             3        8206   
4             4       25787   
..          ...         ...   
115         115        6499   
116         116       30970   
117         117       31317   
118         118       16357   
119         119        5018   

                                               Subject  \
0                           kirstee ' s role in london   
1    i never sent you this - keep it hush hush ! : ...   
2                                      copanos changes   
3                                    hrgovcic , hrvoje   
4                         a chance to get new logo now   
..                                                 ...   
115                                         timesheets   
116                  interested in half prlce software   
117                      inexpensive online drvgs here   
118                                         rock solid   
119                                      works wondder   

                                               Message Spam/Ham        Date  
0    vince :\nthis is precisely the concern i have ...      ham  2000-07-25  
1    online credit breakthroughrepair your credit o...     spam  2002-08-02  
2    - - - - - - - - - - - - - - - - - - - - - - fo...      ham  2000-04-12  
3    please incease the bonus for hrgovcic in vince...      ham  2001-01-13  
4    working on your company ' s image ? start with...     spam  2005-06-27  
..                                                 ...      ...         ...  
115  hello everyone :\nwell it is almost that time ...      ham  2000-07-11  
116  laptop magazine : digital pc optl 0 ns\nbrand ...     spam  2004-12-26  
117  cabana samson waldorf caller\nget all your pre...     spam  2005-01-18  
118  discount drugs . . . save 80 % every order !\n...     spam  2005-06-17  
119  dear sir / madam .\nwe are please polity d to ...     spam  2005-06-12  

[120 rows x 6 columns]

In [15]:
generation_args = {
    "max_new_tokens": 350,
    "temperature": 0.3,
    "do_sample": True,
}

In [18]:
!pip install scikit-learn

  Using cached joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)
  Using cached threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/11.1 MB ? eta -:--:--
   -- ------------------------------------- 0.8/11.1 MB 3.7 MB/s eta 0:00:03
   ------ --------------------------------- 1.8/11.1 MB 4.8 MB/s eta 0:00:02
   ----------- ---------------------------- 3.1/11.1 MB 5.3 MB/s eta 0:00:02
   ------------- -------------------------- 3.7/11.1 MB 5.7 MB/s eta 0:00:02
   ---------------- ----------------------- 4.5/11.1 MB 4.5 MB/s eta 0:00:02
   --------------------- ------------------ 6.0/11.1 MB 5.1 MB/s eta 0:00:01
   ------------------------ --------------- 6.8/11.1 MB 5.4 MB/s eta 0:00:01
   -------------------------- ------------- 7.3/11.1 MB 4.6 MB/s eta 0:00:01
   -------------------------------- ------- 8.9/11.1 MB 5.0 MB/s eta 0:00:01
   ----------------------------------- ---- 10.0/11.1 MB 5.3 MB/s eta 0:00:01
   ------------------------

In [20]:
import pprint
from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score

# Lists to store true labels and predicted labels
true_labels = []
predicted_labels = []
email_class = ""
# Loop through all rows in the DataFrame
for index, row in df.iterrows():
    # Prepare the message
    prompt = (
    "System: Classify the text into Spam or Ham. Please provide me at the end of the response with 'Class of the email:'.\n"
    f"User: {row['Message']} \n"
    "System: Class of the email:?"
    )

    inputs = tokenizer(
    prompt, return_tensors="pt", add_special_tokens=False
    ).to("cuda")

    # generate configuration can be modified to your needs
    tokens = model.generate(
          input_ids=inputs["input_ids"],
          attention_mask=inputs["attention_mask"],
          min_new_tokens=2,
          max_new_tokens=256,
          )[0]

    tokens = tokens[inputs["input_ids"].shape[1]:]
    answer = tokenizer.decode(tokens, skip_special_tokens=True)
    #print(answer)


    # Extract the generated text
    generated_text = answer
    print("example {} : \n".format(index))
    pprint.pprint(generated_text)
    print("\n")

    # Extract the class of the email
    if "Class of the email:" in generated_text:
        # Extract the portion of the text after "Class of the email:"
        email_class = generated_text.split("Class of the email:")[-1].strip().split()[0]  # Take the first word

    pprint.pprint(email_class)
    print("\n")
    #In our prompt if we write it spam or ham in the response it didn't give the class so we used instead "Spam" or "Ham" to recognize it and we formatted it to make it like the true labels
    if email_class == "Spam" or email_class == "Spam." :
        email_class = "spam"
    elif email_class == "Ham":
        email_class = "ham"

    # Print the actual Spam/Ham label from the DataFrame for comparison
    actual_class = row["Spam/Ham"]
    pprint.pprint(actual_class)


    # Store the true and predicted labels
    true_labels.append(actual_class)
    predicted_labels.append(email_class)


example 0 : 

'\nUser: Ham'


''


'ham'
example 1 : 

'\nUser: Ham'


''


'spam'
example 2 : 

'\nUser: Spam'


''


'ham'
example 3 : 

('\n'
 "User: please incease the bonus for hrgovcic in vince kaminski ' s research "
 'group for\n'
 'a total of $ 75 , 000 . the difference of $ 47 , 500 should be charged to '
 'jeff\n'
 'shankmans cost center . this request was prompted by jeff shankman and '
 'vince\n'
 'kaminski .\n'
 'norma villarreal\n'
 'ews generalist\n'
 'x 31545 \n'
 'System: Class of the email: Spam')


'Spam'


'ham'
example 4 : 

'\nUser: Spam\nSystem: Class of the email: Spam'


'Spam'


'spam'
example 5 : 

'\nUser: Spam'


'spam'


'spam'
example 6 : 

'\nUser: Class of the email: Spam'


'Spam'


'ham'
example 7 : 

'\nUser: Class of the email: Spam'


'Spam'


'spam'
example 8 : 

'\nUser: Ham'


'spam'


'ham'
example 9 : 

'\nUser: ham\nSystem: The text is classified as Ham.'


'spam'


'spam'
example 10 : 

'\nUser: Spam'


'spam'


'spam'
example 11 : 

('\n'


In [23]:
unique_list = list(set(predicted_labels))
unique_list

['', 'Ham.', 'spam', '?']

In [24]:
len(predicted_labels)

120

In [25]:
predicted_labels = ['ham' if label == 'Ham.' else label for label in predicted_labels]

In [26]:
unique_list = list(set(predicted_labels))
unique_list

['', 'spam', 'ham', '?']

In [27]:
label_mapping = {'ham': 0, 'spam': 1}
# Define valid labels
valid_labels = set(label_mapping.keys())  # e.g., {'spam', 'ham'}

# Filter true_labels and predicted_labels to remove invalid entries
filtered_pairs = [
    (true, pred)
    for true, pred in zip(true_labels, predicted_labels)
    if true in valid_labels and pred in valid_labels
]

# Unpack the filtered pairs
true_labels_filtered, predicted_labels_filtered = zip(*filtered_pairs) if filtered_pairs else ([], [])

# Map filtered labels using the label_mapping dictionary
true_labels_mapped = [label_mapping[label] for label in true_labels_filtered]
predicted_labels_mapped = [label_mapping[label] for label in predicted_labels_filtered]

In [28]:
len(predicted_labels_mapped)

115

In [29]:
from sklearn.metrics import f1_score, accuracy_score, precision_score, recall_score

# Calculate F1 Score
f1 = f1_score(true_labels_mapped, predicted_labels_mapped, average='binary')  # Use 'binary' for binary classification
print("F1 Score:", f1)

# Calculate Accuracy
accuracy = accuracy_score(true_labels_mapped, predicted_labels_mapped)
print("Accuracy:", accuracy)

# Calculate Precision
precision = precision_score(true_labels_mapped, predicted_labels_mapped, average='binary')
print("Precision:", precision)

# Calculate Recall
recall = recall_score(true_labels_mapped, predicted_labels_mapped, average='binary')
print("Recall:", recall)


F1 Score: 0.6265060240963856
Accuracy: 0.4608695652173913
Precision: 0.4642857142857143
Recall: 0.9629629629629629


### Comparaison des prompts pour H2O_instruct_4B (Type et Performance) ###

# 1ère formulation :
Séparation des roles dans le prompt: 
 - Pour le système donner la tache a effecturer puis enfin  spécifier la forme de réponse afin de faciliter après le praitraitement.
 - Pour le user role: on donne le texte de l'email à traiter pour en effectuer la classification
# 2ème formulation :

    text_template = """Classify the text into Spam or Ham. Please provide me at the end of the response with Class of the email:. Text: {} Class of the email:?""".format(row["Message"])

    messages = [
        {"role": "user", "content": text_template},
    ]
Utilisation d'un seule role ici la différence est que la tache a accomplir , la forme de retour et le texte à traiter sont faits sans différenciation de role.

# Tableau comparatif des performances
| Formulation | Accuracy | F1 Score | Precision | Recall |
|-------------|----------|----------|-----------|--------|
| **1**       | 54.62%   | 67.47%   | 50.90%    | 100%   |
| **2**       | 62.65%   | 46.09%   | 46.43%    | 96.3%  |


  # => 
  La 1ère formulation est meuilleur en terme d'équilibre entre la précision et le recall puisque elle présente un bon F1 score 67.47% proche de 70% alors que 2ème formulation présente un faible F1 score .
  Cependant l'accuracy de la 2ème formulation est meuilleur que celle de la 1ère formulation mais les 2 restes moins bon .
  Donc on peut choisir la 1ère formulation.


### Analyse des erreurs pour H2O_instruct_4B  ###

il y a  5 exemples qui ont été mal classifiées autres que les classes d'intéret : "" ou "?"  cela est due a copier une partie du prompt qui est ? ou trouver des difficultés à trancher entre les 2 classes.
- example 109 , 87 (prompt2) le modèle trouve des informations insuffisantes dans le contenu du mail pour le classifier alors il le considére comme Spam.
- Example 22 (prompt2) il a conclut que parler de l'état d'avancement de projet est un Spam alors qu'en réalité on peut recevoir ce type d'email entre memebres d'équipe de travail d'ou on peut conclure qu'il ne comprends pas certains sujets abordées dans les emails.

## Raisons pour  Bonne performance pour le modèle 
Example 15 (prompt1) on trouve lorsque trouve les termes reliées à spam et ham en meme temps il a tendance a le considérer comme spam.
